src/Microsoft.AspNet.Hosting/Internal/WebApplication.cs

```diff
@@ -83,11 +80,43 @@ public IServiceProvider ApplicationServices
             }
         }
 
-        public virtual IApplication Start()
+        public IFeatureCollection ServerFeatures
+        {
+            get
+            {
+                return Server?.Features;
+            }
+        }
+
+        public IApplicationLifetime Lifetime
+        {
+            get
+            {
+                return _applicationLifetime;
+            }
+        }
+
+        public IHostingEnvironment HostingEnvironment
+        {
+            get
+            {
+                return Services.GetRequiredService<IHostingEnvironment>();
+            }
+        }
+
+        public ILoggerFactory LoggerFactory
+        {
+            get
+            {
+                return Services.GetRequiredService<ILoggerFactory>();
+            }
+        }
+
+        public virtual IDisposable Start()
         {
             var application = BuildApplication();
```

Comment:
Wasn't BuildApplication going to move to the constructor? Otherwise ServerFeatures isn't available before Start.